client/dashboard/sites/views.ts

```diff
@@ -80,31 +101,122 @@ export function getView( {
 	user,
 	isAutomattician,
 	isRestoringAccount,
-	viewOptions,
 	viewPreferences,
+	viewSearchParams,
 }: {
 	user: User;
 	isAutomattician: boolean;
 	isRestoringAccount: boolean;
-	viewOptions: Partial< ViewTable | ViewGrid >;
-	viewPreferences?: Partial< View >;
+	viewPreferences?: ViewPreferences;
+	viewSearchParams: ViewSearchParams;
 } ): {
-	defaultView: View;
-	view: View;
+	defaultView: SitesView;
+	view: SitesView;
 } {
 	const defaultView = getDefaultView( {
 		user,
 		isAutomattician,
 		isRestoringAccount,
 	} );
 
-	const type = viewOptions.type || viewPreferences?.type || defaultView.type;
+	const type = viewSearchParams.type || viewPreferences?.type || defaultView.type;
+
 	const view = {
 		...defaultView,
 		...DEFAULT_LAYOUTS[ type ],
+		...DEFAULT_LAYOUT_FIELDS[ type ],
 		...viewPreferences,
-		...Object.fromEntries( Object.entries( viewOptions ).filter( ( [ , v ] ) => v !== undefined ) ),
-	} as View;
+		...viewSearchParams,
+	} as SitesView;
+
+	return {
+		defaultView,
+		view,
+	};
+}
+
+// Returns the updated view preferences and search params, after `view` has changed to `nextView`.
+// The updated view preferences are produced by merging `nextView` into `viewPreferences`.
+// Finally, the updated view preferences and search params are filtered to remove any default values,
+// so that the remaining values reflect the user's explicit intent.
+export function mergeViews( {
+	defaultView,
+	view,
+	viewPreferences,
+	nextView,
+}: {
+	defaultView: SitesView;
+	view: SitesView;
+	viewPreferences?: ViewPreferences;
+	nextView: SitesView;
+} ): {
+	updatedViewPreferences: ViewPreferences;
+	updatedViewSearchParams: ViewSearchParams;
+} {
+	const nextType = nextView.type;
+
+	let updatedView = nextView;
+
+	if ( nextType !== view.type ) {
+		updatedView = {
+			...updatedView,
+
+			// If the type is changed, we restore the default layout for that type.
+			...DEFAULT_LAYOUTS[ nextType ],
+		} as SitesView;
+
+		if ( ! viewPreferences?.fields ) {
+			updatedView = {
+				...updatedView,
+
+				// If the type is changed, and the user has never explicitly set custom fields,
+				// we restore the default fields for that type.
+				...DEFAULT_LAYOUT_FIELDS[ nextType ],
```

Comment:
Looks like we can keep it in `DEFAULT_LAYOUTS` and then override the fields if the user has explicitly set custom fields 🤔

Reply:
This is the flash that happens if we keep them in `DEFAULT_LAYOUTS`:


https://github.com/user-attachments/assets/8c7bb48c-f209-4546-a287-3e35d26aa612